Unmatched crm facts

anshuldubela opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
This testcase toggles the mux state from standby to active to standby. Testcase fails because the crm facts before and after are not matching.
<!--
- Due to unmatched_crm_facts minimum length which should be 2, the assert was failing.
- Moved the crm_facts next to sleep post observing the other implementations. 
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305

### Approach
#### What is the motivation for this PR?
Testcase fails because the crm facts before and after are not matching.
#### How did you do it?
By Verifying different approaches with crm facts collection.
#### How did you verify/test it?
Verified on T0 Hardware.
#### Any platform specific information?
NA
#### Supported testbed topology if it's a new test case?
NA
### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

anshuldubela committed:
MIGSOFTWAR-8454 Unmatched CRM facts

Signed-off-by: Anshul Dubela <andubela@cisco.com>

anshuldubela committed:
Merge branch 'sonic-net:master' into master

anshuldubela committed:
MIGSOFTWAR-8454 Unmatched CRM facts

Signed-off-by: Anshul Dubela <andubela@cisco.com>

anshuldubela committed:
Merge branch 'master' of github.com:anshuldubela/sonic-mgmt

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

"tests/dualtor/test_standby_tor_upstream_mux_toggle.py:89:121: E501 line too long (123 > 120 characters)"

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

"tests/dualtor/test_standby_tor_upstream_mux_toggle.py:89:121: E501 line too long (123 > 120 characters)"

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

    # For Cisco-8000 devices, hardware counters are statistical-based with +/- 1 entry tolerance.
    # Hence, the available counters may not increase as per initial value for multiple facts collected.

anshuldubela committed:
Update utilities.py

        # For Cisco-8000 devices, hardware counters are statistical-based with +/- 1 entry tolerance.
        # Hence, the available counter may not increase as per initial value.

anshuldubela committed:
Update utilities.py

Removed the extra line.

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

Fix below pre-commit error:
tests/dualtor/test_standby_tor_upstream_mux_toggle.py:95:39: E225 missing whitespace around operator

anshuldubela committed:
Update utilities.py

Fixed Below pre-commit issue:
trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing tests/common/utilities.py

anshuldubela committed:
Merge branch 'sonic-net:master' into master

anshuldubela committed:
Merge branch 'sonic-net:master' into master

anshuldubela committed:
Merge branch 'sonic-net:master' into master

anshuldubela committed:
Merge branch 'sonic-net:master' into master

anshuldubela committed:
Merge branch 'sonic-net:master' into master

anshuldubela committed:
Unmatched CRM facts Rev

Signed-off-by: Anshul Dubela <andubela@cisco.com>

anshuldubela committed:
Merge pull request #3 from anshuldubela/andubela/MIGSOFTWAR-8454-FIX-REV

Rev 8454 Unmatched CRM facts rev

anshuldubela committed:
Unmatched_CRM_facts

Signed-off-by: Anshul Dubela <andubela@cisco.com>

lolyu commented on tests/dualtor/test_standby_tor_upstream_mux_toggle.py:
```diff
@@ -20,7 +20,7 @@
 ]
 
-PAUSE_TIME = 10
+PAUSE_TIME = 90
```
Why do we need to increase the pause time?

bpar9 replied:
the pause needs to take into effect only for cisco devices and time needs to be checked in a loop vs. always waiting for 90

bpar9 commented on tests/common/utilities.py:
```diff
@@ -488,7 +488,10 @@ def compare_crm_facts(left, right):
         rv = right['resources'][k]
-        if lv['available'] != rv['available'] or lv['used'] != rv['used']:
+
+        # For Cisco-8000 devices, hardware counters are statistical-based with +/- 1 entry tolerance.
```
this will remove the check for all vendors and add only cisco specific check for all. This should not be done this way. You need to add the check to only take into effect when running on Cisco specific devices

bpar9 commented on tests/common/utilities.py:
```diff
@@ -510,7 +513,10 @@ def compare_crm_facts(left, right):
         rv = right_acl_group[k]
-        if lv['available'] != rv['available'] or lv['used'] != rv['used']:
+
+        # For Cisco-8000 devices, hardware counters are statistical-based with +/- 1 entry tolerance.
```
same as above comment, this check needs to be added to take into effect only while running against cisco specific devices

anshuldubela committed:
Unmatched_CRM_facts added cisco checks

Signed-off-by: Anshul Dubela <andubela@cisco.com>

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/utilities.py:493:28: F821 undefined name 'duthost'<br/>tests/common/utilities.py:522:28: F821 undefined name 'duthost'<br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:24:20: F405 'duthost' may be undefined, or defined from star imports: tests.common.dualtor.dual_tor_mock<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

wsycqyz commented:
@anshuldubela The pre-commit check is a mandatory check. Please help fix accordingly.

anshuldubela committed:
Unmatched_CRM_facts added cisco checks

Signed-off-by: Anshul Dubela <andubela@cisco.com>

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:24:20: F405 'duthost' may be undefined, or defined from star imports: tests.common.dualtor.dual_tor_mock<br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:92:45: F405 'duthost' may be undefined, or defined from star imports: tests.common.dualtor.dual_tor_mock<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

anshuldubela committed:
Unmatched_CRM_facts added cisco checks

Signed-off-by: Anshul Dubela <andubela@cisco.com>

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

Removed the is_cisco_device usage and added the check by duthost.

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:42:8: F405 'duthosts' may be undefined, or defined from star imports: tests.common.dualtor.dual_tor_mock<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

anshuldubela committed:
Update utilities.py

Removed the is_cisco_device check and added the duthost check.

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/utilities.py:492:12: F821 undefined name 'duthosts'<br/>tests/common/utilities.py:521:12: F821 undefined name 'duthosts'<br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:42:8: F405 'duthosts' may be undefined, or defined from star imports: tests.common.dualtor.dual_tor_mock<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

duthosts to duthost

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/utilities.py:492:12: F821 undefined name 'duthosts'<br/>tests/common/utilities.py:521:12: F821 undefined name 'duthosts'<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

anshuldubela committed:
Update utilities.py

duthosts to duthost

bpar9 commented on tests/common/utilities.py:
```diff
@@ -473,7 +473,7 @@ def is_ipv4_address(ip_address):
 
 
-def compare_crm_facts(left, right):
+def compare_crm_facts(duthost, left, right):
```
you cannot specify the arguments this way as other scripts are also utilizing the same proc. Need the arguments to be optional for backward compatibility

bpar9 commented on tests/common/utilities.py:
```diff
@@ -488,8 +488,15 @@ def compare_crm_facts(left, right):
+        if duthost.facts['platform_type'] == "cisco-8000":
+            # For Cisco-8000 devices, hardware counters are statistical-based with +/- 1 entry tolerance.
+            # Hence, the available counter may not increase as per initial value.
+            if abs(int(lv['available']) - int(rv['available'])) > 1 or abs(int(lv['used']) - int(rv['used'])) > 1:
```
another option is, instead of adding the check here, get_crm_facts can be collected and the comparison can be done within the calling script itself. Or go with the optional arguments method

bpar9 commented on tests/dualtor/test_standby_tor_upstream_mux_toggle.py:
```diff
@@ -79,7 +91,12 @@ def test_standby_tor_upstream_mux_toggle(
+    logging.debug(
+      "unmatched_crm_facts: %s len: %s" % (unmatched_crm_facts, len(unmatched_crm_facts)))
+    # Verify packets are not go up
+    pt_assert(len(unmatched_crm_facts) <= 5, 'Unmatched CRM facts: {}'
```
this change cannot be made for all vendors. It has to be cisco specific

anshuldubela committed:
Unmatched_CRM_facts review comments

Signed-off-by: Anshul Dubela <andubela@cisco.com>

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:103:17: E128 continuation line under-indented for visual indent<br/>tests/dualtor/test_standby_tor_upstream_mux_toggle.py:108:17: E128 continuation line under-indented for visual indent<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

anshuldubela committed:
Update test_standby_tor_upstream_mux_toggle.py

tests/dualtor/test_standby_tor_upstream_mux_toggle.py:103:17: E128 continuation line under-indented for visual indent
tests/dualtor/test_standby_tor_upstream_mux_toggle.py:108:17: E128 continuation line under-indented for visual indent

jamesan47 commented:
Hi @bpar9 , could you assist Ansul with the merge conflict?

bpar9 commented:
> Hi @bpar9 , could you assist Ansul with the merge conflict?

@jamesan47 , the conflicts are in utilities.py, Anshul needs to address those accommodating for newer changes  in that file

anshuldubela committed:
Merge branch 'master' into anshuldubela/Unmatched_CRM_facts

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/utilities.py:479:1: E302 expected 2 blank lines, found 1<br/>tests/common/utilities.py:491:1: E302 expected 2 blank lines, found 1<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

anshuldubela committed:
Update utilities.py

tests/common/utilities.py:479:1: E302 expected 2 blank lines, found 1
tests/common/utilities.py:491:1: E302 expected 2 blank lines, found 1
